x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/TransportTermEnumAction.java

```diff
@@ -0,0 +1,595 @@
+            shardsIts = shards(clusterState, request, concreteIndices);
+            expectedOps = shardsIts.size();
+
+            shardsResponses = new AtomicReferenceArray<>(expectedOps);
+        }
+
+        public void start() {
+            if (shardsIts.size() == 0) {
+                // no shards
+                try {
+                    listener.onResponse(newResponse(request, new AtomicReferenceArray(0), clusterState, false));
+                } catch (Exception e) {
+                    listener.onFailure(e);
+                }
+                // TODO or remove above try and instead just call finishHim() here? Helps keep return logic consistent
+                return;
+            }
+            // count the local operations, and perform the non local ones
+            int shardIndex = -1;
+            for (final ShardIterator shardIt : shardsIts) {
+                if (checkForEarlyFinish()) {
+                    return;
+                }
+                shardIndex++;
+                final ShardRouting shard = shardIt.nextOrNull();
+                if (shard != null) {
+                    performOperation(shardIt, shard, shardIndex);
+                } else {
+                    // really, no shards active in this group
+                    onOperation(null, shardIt, shardIndex, new NoShardAvailableActionException(shardIt.shardId()));
+                }
+            }
+        }
+        
+        // Returns true if we exited with a response to the caller.
+        boolean checkForEarlyFinish() {
+            long now = System.currentTimeMillis();
+            if ( (now - task.getStartTime()) > request.timeout().getMillis() ) {
+                finishHim(true);
+                return true;
+            }
+            return false;
+        }
+
+        protected void performOperation(final ShardIterator shardIt, final ShardRouting shard, final int shardIndex) {
+            if (shard == null) {
+                // no more active shards... (we should not really get here, just safety)
+                onOperation(null, shardIt, shardIndex, new NoShardAvailableActionException(shardIt.shardId()));
+            } else {
+                try {
+                    //TODO pass through a reduced timeout (the original time limit, minus whatever we may have
+                    // spent already getting to this point.
+                    final ShardTermEnumRequest shardRequest = newShardRequest(shardIt.size(), shard, request);
+                    shardRequest.setParentTask(clusterService.localNode().getId(), task.getId());
+                    DiscoveryNode node = nodes.get(shard.currentNodeId());
+                    if (node == null) {
+                        // no node connected, act as failure
+                        onOperation(shard, shardIt, shardIndex, new NoShardAvailableActionException(shardIt.shardId()));
+                    } else {
```

Comment:
The timeout could be checked here to save some round trip and could avoid accumulating requests on the node's thread pool ?